3D GEQ

netmindz opened the pull request:
PR to pull in @troyhacks new 3D GEQ effect

netmindz committed:
Add @TroyHacks 3D GEQ

netmindz committed:
Add @TroyHacks 3D GEQ

netmindz committed:
;

netmindz committed:
3D GEQ - make front brightness a control

netmindz commented on wled00/FX.cpp:
```diff
@@ -8355,6 +8355,138 @@ uint16_t mode_2Dwavingcell() {
+        }
+      }
+
+    if(frontBrightness >= 50) { // TODO: other values too? not sure exactly when we want the border
```
Is this right for the border?

netmindz replied:
@troyhacks ?

troyhacks replied:
I don't know if I even like the borders. They were kinda more a demo hack.

troyhacks replied:
Borders have been removed. I had a strong aesthetic dislike of them during testing.

netmindz commented on wled00/FX.cpp:
```diff
@@ -8355,6 +8355,138 @@ uint16_t mode_2Dwavingcell() {
+      else if(frontBrightness >= 50) {
+        // Faded fronts, assumes border added later.
+        for (int x=linex+1; x<linex+(cols/16)-1;x++) { 
+          SEGMENT.drawLine(x,rows-2,x,rows-heights[i]-2,color_fade(ledColor,64,true)); // front fill
```
I have allowed a range of 200 for different fade amounts, do you know what would be a good range to map this to @troyhacks ?

netmindz replied:
using the full range for now so you can move from the no fill smoothly to the fully filled

netmindz commented on wled00/FX.cpp:
```diff
@@ -8355,6 +8355,138 @@ uint16_t mode_2Dwavingcell() {
+
+  // Author: @TroyHacks
+
+  static uint16_t projector;
+  static uint16_t projector_dir = 1;
```
Moved to be inside the scope of the method to try and prevent yet another global variable in WLED

MoonModules replied:
Variables which persists between loops should be done like this

<img width="670" alt="image" src="https://github.com/MoonModules/WLED/assets/91013628/4b84cf03-1a2f-49dd-a2f4-b41226d7dcff">

netmindz replied:
Updated

netmindz commented on wled00/FX.cpp:
```diff
@@ -8355,6 +8355,138 @@ uint16_t mode_2Dwavingcell() {
+    SEGMENT.fadeToBlackBy(SEGMENT.speed);
+  }
+
+  const int NUM_BANDS = 16; // map(SEGMENT.custom1, 0, 255, 1, 16);
```
I'm guessing if we swap back to being configurable then we would also need to change a few of the current magic values in the code? probably most of the 16s with NUM_BANDS

netmindz replied:
I think I have updated it all correctly, just one comment about FFT band selection

netmindz committed:
3D GEQ - customisable number of bands

netmindz committed:
3D GEQ - variable front brightness

netmindz committed:
3D GEQ - customisable number of bands

netmindz commented:
Should it perhaps be GEQ 3D rather than 3D GEQ so it sits next to the other in the list?

netmindz commented on wled00/FX.cpp:
```diff
@@ -8355,6 +8355,137 @@ uint16_t mode_2Dwavingcell() {
+    SEGMENT.setUpLeds(); // WLEDMM use lossless getPixelColor()
+    SEGMENT.fill(BLACK);
+  } else {
+    projector += projector_dir;
+    if (projector == SEGMENT.virtualWidth()) projector_dir = -1;
+    if (projector == 0) projector_dir = 1;
```
Could - add speed control here so you can move slower?

troyhacks replied:
yeah, that's totally possible. Just scale the unit of increment.

netmindz replied:
To make it slower I would surely need to only update every X frames rather than change the increment as it's an int?

troyhacks replied:
Done!

troyhacks commented:
> Should it perhaps be GEQ 3D rather than 3D GEQ so it sits next to the other in the list?

I was thinking the same thing.

netmindz committed:
3D GEQ - fade speed

MoonModules commented on wled00/FX.cpp:
```diff
@@ -8355,6 +8355,137 @@ uint16_t mode_2Dwavingcell() {
 }
 static const char _data_FX_MODE_2DWAVINGCELL[] PROGMEM = "Waving Cell@!,,Amplitude 1,Amplitude 2,Amplitude 3;;!;2";
 
+uint16_t mode_3DGEQ(void) {
```
Maybe better to call it laserGEQ or something and not use 3D as in StarLight 3D means an effect in 3 dimensions?

netmindz replied:
Changed to GEQ Laser  - so it sits next to the other in the list

netmindz committed:
Rename 3D GEQ to GEQ Laser

netmindz committed:
GEQ Laser use SEGMENT.data

troyhacks committed:
Bugfixed GEQ 3D

Bug fixes after lots of testing, better settings, etc.

Big thanks to @netmindz  for restructuring the code.

Add rough distance stop to Segment::drawLine()

troyhacks committed:
Merge branch 'mdev' into 3DGEQ

troyhacks committed:
Seems someone improved drawline() - fixed for this

troyhacks committed:
Don't use new soft drawline() option

troyhacks committed:
Oopsy

troyhacks reviewed (approved):
Works for me!

I added a lot of rendering bugfixes to work with all the new various sliders so it always behaves as expected.

"Speed" is implemented with the fastest setting as 1 pixel shift per frame, and that goes down to 1 pixel shift every 10 frames at the lowest end of the slider.

Thanks to @netmindz for restructuring so it uses proper WLED functionality. 

I've tested this down to 16x16 and you can get something that looks pretty good at that resolution if you adjust the sliders, reduce the bars, etc. 

I'm not overly happy with the perspective when the horizon is moved and the distance is lowered - I feel like that needs some sort of "depth" change so it's not coming to a point when the projector point is right over the bar.

...but that's for a later fix. 😄

netmindz commented:
I'm in favour of more controls, however the dmx effect mode personality only supports the speed and intensity options, so are we happy we are mapping the two most useful controls to the right thing? Personally fade speed is the least useful to me, for it to look most like a laser and appear most responsive that is with zero fade. Whereas I'm highly likely to want to change the number of bands and would not be able to in Effect Mode.

ewoudwijma commented:
Why is dmx only supporting 2 controls? Dmx fixtures have often easily 10+ channels. Shouldn't WLED support that as well?

netmindz commented:
You could have 512 used by a singular fixture if you wanted, it's just how WLED has chosen which channels can be assigned to what purpose 


https://kno.wled.ge/interfaces/e1.31-dmx/#effect

I'm guessing only speed and intensity existed when the mapping was done. We could add a new mode (personality) where all possibly effect parameters are exposed to DMX, but that would be it's own feature and PR

softhack007 commented on wled00/FX.cpp:
```diff
@@ -8355,6 +8355,155 @@ uint16_t mode_2Dwavingcell() {
+  uint8_t heights[NUM_BANDS] = { 0 };
+
+  for (int i=0; i<NUM_BANDS; i++) {
+    heights[i] = map8(fftResult[i],0,rows*0.85);
```
* What is `map8()`?
* I think this loop can be optimized out (will make a proposal later)

softhack007 replied:
Nah, let's leave this like it is.

softhack007 commented on wled00/FX.cpp:
```diff
@@ -8355,6 +8355,155 @@ uint16_t mode_2Dwavingcell() {
+    uint16_t colorIndex = map(cols/NUM_BANDS*i, 0, cols-1, 0, 255);
+    uint32_t ledColor = SEGMENT.color_from_palette(colorIndex, false, PALETTE_SOLID_WRAP, 0);
+
+    int linex = i*(cols/NUM_BANDS);
```
Should be "(i * cols) / NUM_BANDS" to improve accuracy.

softhack007 replied:
Update: tried the idea - it cases better distribution over available horizontal pixels. But also its creating gaps because some bars are a pixel wider than others. Maybe something for later.

softhack007 commented on wled00/FX.cpp:
```diff
@@ -8355,6 +8355,155 @@ uint16_t mode_2Dwavingcell() {
+
+  for (int i=(NUM_BANDS - 1); i>split; i--) { // paint left vertical faces and top - RIGHT to LEFT
+
+    uint16_t colorIndex = map(cols/NUM_BANDS*i, 0, cols-1, 0, 255);
```
(Not 100% sure) I think this is the same as "(i * cols) / NUM_BANDS" which is more accurate

softhack007 commented:
I see a few minor optimizations for accuracy, but spotted nothing yet that could drastically improve speed. 
I'll take a close look tomorrow or monday.

softhack007 committed:
(experimental) setPixelColorXY_fast speedup by setPixelColorXY_fast

* adding a _fast_ variant of SetPixelColorXY, that does not perform any error checking.
* drawLine uses the fast setPixelColor variant (bresenham algo only)
* for TESTING: segment option "reverse" switches back to the "original slow" code.
surely needs some more optimization and improvements.

First test on GEQ 3D shows 10%-30% speedup

softhack007 commented:
@netmindz @troyhacks I've pushed a first shot for faster DrawLine, using an optimized setPixelColorXY_fast that does not repeat error checking and cut a few more corners.
Currently only supports GEQ 3D, but could be later extended to other "draw loops". 

Can give checking this give higher framerates in your setups? I'd expect around 10%-30% speedup.

troyhacks commented:
> @netmindz @troyhacks I've pushed a first shot for faster DrawLine, using an optimized setPixelColorXY_fast that does not repeat error checking and cuts a few more corners. Currently only supports GEQ 3D, but could be later extended to other "draw loops".
> 
> Can you check if this produces better framerates in your setups? I'd expect around 10%-30% speedup.

This is good for 2-3 FPS faster at around 20 FPS before.

softhack007 committed:
more sPC optimizations

* made setPixelColorXY_fast private
* optimized Segment::fill() to use setPixelColorXY_fast
* "bar" 1D expand optimization

troyhacks committed:
Optimized side-wall drawing

Only draw what will be visible.

This is a pretty solid speed increase when the adjacent bar is close in height, or if it's taller and would occlude it anyway, it'll be skipped.

troyhacks commented:
Added optimized side-wall drawing that will only draw what will be visible, rather than the entire height.

This is a pretty solid speed increase when the adjacent bar is close in height - or if it's taller and would  be occluded anyway, it'll be skipped.

netmindz commented:
> Added optimized side-wall drawing that will only draw what will be visible, rather than the entire height.
> 
> This is a pretty solid speed increase when the adjacent bar is close in height - or if it's taller and would  be occluded anyway, it'll be skipped.

What impact did this have on speed?

softhack007 committed:
Merge branch 'mdev' into 3DGEQ

softhack007 committed:
prevent array bounds violation

+ don't use dynamic array for heights[] - dynamic arrays are using malloc() -> heap frag.

softhack007 committed:
loop optimizations

* made cols/rows integer, to avoid that "cols-x" wraps around
* pre-compute some expressions that don't depend on loop counters

--> wins us 1-4 fps on my 72x72 test setup

troyhacks commented:
> > What impact did this have on speed?
> 
> on my test setup: 20fps --> 30-35fps.

![image](https://github.com/user-attachments/assets/b9697422-66f6-43a8-8466-b044473a8d08)

Before it would have drawn a sidewall for every bar except the one under the projector, from the bottom to the height (like in the red circle)

...but now it only draws the green areas because those sidewalls are actually visible.

troyhacks committed:
Borders checkbox for @netmindz

Netmindz likes the borders. Now a check-box to enable.

troyhacks committed:
Borders checkbox

troyhacks commented:
Added borders back as a checkbox for @netmindz  😊

softhack007 commented:
@troyhacks my fault, its there. I just had to move the "front fill" slider to left side to see the boxes.

![image](https://github.com/user-attachments/assets/47fc77f3-28b5-4e4b-9a18-82698392986c)

softhack007 committed:
speedup for pinwheel mapping

fast sPC() in pinwheel mapping -> ~10% speedup

troyhacks commented:
> @troyhacks my fault, its there. I just had to move the "front fill" slider to left side to see the boxes.
> 
> ![image](https://github.com/user-attachments/assets/47fc77f3-28b5-4e4b-9a18-82698392986c)
> 

It should work for values below "full" down to black-out. Full front fill is the same color.

I guess I could make them not draw when the front face fill is 100%, but that's only a very minor speedup and there's a checkbox for that.

softhack007 committed:
GEQ speedup

use drawLine() to draw a complete bar at once. Only possible if _not_ doing colored bars.

softhack007 committed:
better handling of perspective

drawLine parameter "depth" shortens a line by a fraction - useful for VP perspective (vanishing point)

softhack007 commented:
added better handling of depth (distance) in DrawLine - could be a bit improved with proper rounding...

30% depth <img width="45%" src="https://github.com/user-attachments/assets/8c11a866-42bb-4112-b574-c84bdcfd6d51">

100% depth <img width="45%" src="https://github.com/user-attachments/assets/a423f99d-da36-499b-ae08-55ec4a7a2e09">

softhack007 committed:
better rounding

softhack007 committed:
soft light option

* bugfix for Borders (did not work with <16 bands)

* new option for soft light: when checked, the left & right edges of top lights are drawn with anti-aliasing on. Costs us 1-2 fps, so I made it optional.

softhack007 committed:
workaround for "cannot read properties of undefined. reading 'split'

hmpf ... looks like somewhere deep in WLED, the FX name string gets limited to 127 chars....

softhack007 commented:
new feature: "Soft Light" - draws the light rays with Anti-Aliasing enabled. Costs 1-2 fps, so I made it optional.

![image](https://github.com/user-attachments/assets/337ab161-e816-41a8-9078-c430cdc261c1)
